feat(client): initial support for HTTP/2

mlalic opened the pull request:
This PR adds initial support for HTTP/2 requests.

It introduces a new pair of traits: `HttpMessage` and `Protocol`.

The `HttpMessage` trait provides an API that abstracts the sending and receiving of HTTP messages away from the concrete implementation of the underlying protocol.

The `Protocol` trait provides an API for constructing new `HttpMessage`s.

The `Request` and `Response` structs are refactored to internally use a `Box<HttpMessage>` for their HTTP communication requirements.

The `Client` has a new public method: `with_protocol`. This provides an API for plugging in a particular `Protocol` implementation that will then be used for constructing the `HttpMessage` instances that are passed to the `Request` struct constructor.

The implementation of this API (i.e. the two traits) for HTTP/2 is provided in the `hyper::http2` module.

Therefore, now it is possible to inject an `Http2Protocol` instance when constructing a `hyper::Client` which seamlessly makes it switch to using HTTP/2.

The `Http2Protocol` can use the existing `NetworkConnector` API for constructing the network connection for the underlying communication. However, do note that the current TLS connector does not honor the requirements that the HTTP/2 spec additionally makes for TLS connections (particular cipher suits are banned, applications must use ALPN -- or at least NPN -- for protocol negotiation), so if such a connector is provided, it won't be possible to establish an HTTP/2 connection. The cleartext TCP connectors work as expected.

An example is included in `examples/client_http2.rs` showing how the client can be configured to use HTTP/2.

```
mlalic@skaro:~/src/hyper$ cargo run --example client_http2 -- http://http2bin.org/get?hi=hi
     Running `target/debug/examples/client_http2 http://http2bin.org/get?hi=hi`
Response: 200 OK
Headers:
x-clacks-overhead: GNU Terry Pratchett
access-control-allow-origin: *
:status: 200
content-type: application/json
server: h2o/1.1.1
date: Fri, 29 May 2015 12:29:14 GMT
access-control-allow-credentials: true

{
  "args": {
    "hi": "hi"
  }, 
  "headers": {
    "Connection": "keep-alive", 
    "Host": "http2bin.org,http2bin.org", 
    "Via": "2 http2bin.org"
  }, 
  "origin": "131.159.207.123", 
  "url": "http://http2bin.org,http2bin.org/get?hi=hi"
}
```

# Breaking Changes

The PR introduces only a few minor, however still breaking changes:

- The `hyper::client::Response` struct no longer requires any type parameters.
- The `hyper::Error` public enum has gained a new variant `Http2`

GitCop commented:
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 830ebe2c7c2c02b4ee7c22c3043417cafdac5a55
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 42d4364e4b2c45c8bd8e3d0905f92b26b9b9c727
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: d426c99ed4287d633978e3874a9d4b91465163cd
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: abf43e03ef1cfd6901026799db9be7dfb4bcdf88
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8e61b4351526ce468a887fc30eb0bc62f5a134a0
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 0a3c8e13a8338190c7ce16d5bb2194cf2229ee96
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4eb9786a105c74c2b940954a6f6fbf6da2c44019
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 45f509c9e9e8dd6a382a54c804e77d814488a7ba
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: acd28069931d90775605d730d23515c9945c235c
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: b3395ef688dbbb17cd8f9e3403ed5a654383abc3
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 14c1d60a2af5885cf80bcc3be8b371490f620d66
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 84925772d5411e71a29d4261d6b7dc1bcb6f65b1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ac1fe5467d2b44080c70cdcf6acea36ace4e9e3d
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 11ae013fdb5d2f82cc208a6f6ae4bd0c73fb2842
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 42bcf865f35d44876e618dece0355afa3afd60b4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4d5f293883ce65f0f6dd4cd1f995bf5b17efc590
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 9196bab0529033680b64afc0ac89ddeff24d2c98
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: d8c13fc07d86a13bfc80ddd0ed81df37276a15a1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 121e9ab84c66a2e1c0a0c70fd74503e703127545
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 6ffc23f21d1910073c1c6a24229bb82802049161
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 482e24f834434889feec8b72d2db6ada45072f10
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: e13ce02766cc3c0772ea6ca88d9f4a43f82107c6
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 190101f9f0e2f3c35d38074c542c65f91d088a08
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: f553ffddb8bbc20e0aae694cbd4934fcc32eb0b9
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4ad58bc9753e640d81d78ffc757a4193cee98ee7
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

seanmonstar commented:
So excite.

seanmonstar commented on Cargo.toml:
```diff
@@ -25,6 +25,8 @@ url = "*"
 typeable = "*"
 
+[dependencies.solicit]
+git = "https://github.com/mlalic/solicit"
```
Not on crates.io yet?

mlalic replied:
Yeah, not yet... I wanted to maintain some flexibility, without pushing multiple versions out to crates.io, if it turned out that I needed to do some changes there to make things work better in `hyper` (so far it turned out that I didn't).

Of course it's something to do before finally landing this to make sure `hyper` is pinned to a proper version.

seanmonstar commented on src/client/mod.rs:
```diff
@@ -32,39 +32,120 @@ use status::StatusClass::Redirection;
 /// Clients can handle things such as: redirect policy.
-pub struct Client {
-    connector: Connector,
+pub struct Client<F=Http11RequestFactory> where F: HttpRequestFactory {
```
So as to keep the Client without type parameters, what if we keep a `Box<HttpRequestFactory>` instead?

mlalic replied:
That was my initial idea as well, but after some thought I found it wouldn't work.

You see, the `FreshHttpRequest`, `StreamingHttpRequest`, and `HttpResponse` traits define methods that *move* (i.e. consume) the object they're called on (`start`, `send`). What that means is that you can't call them on trait objects (such as `Box<FreshHttpRequest>`). Therefore, you have to know the exact types of the request and response objects, so even if the factory were boxed up, the `Client` would end up having to be generic over the `FreshHttpRequest` trait (which is what the factory produces). Perhaps there's a different workaround for this, but since the type of the fresh request needs to be known at the point of calling the `start` method, I don't see how we could fully avoid any type parameter to the `Client`.

A different option is to make the request and response methods not consume the object, but that would be a much more invasive change to the codebase and would require a lot more thought about the implications of such a change. I, myself, am not sure whether that'd be a good approach, since at least currently, the requests' "natural" lifecycle is a transformation `Fresh -> Streaming -> Response` and so it is natural that the methods consume/move the object.

reem replied:
There is a hack to get around this where you define another trait that takes `self: Box<Self>` instead of `self`. See `FnBox` in std for an example.

mlalic replied:
Right, that's a nice hack and could be used to circumvent the problem I mentioned.

That said, I'm still not convinced of the benefit of doing it like that, instead of having a generic `Client`. Consider the `send` method on the `RequestBuilder`. In this scenario, its signature needs to change to return `HttpResult<Box<HttpResponse>>` -- since we can't know the actual type of the response. This is also a change to the public API of the client -- and in my opinion a more significant one than introducing a generic type parameter (with a default).

If the `send` method returns a trait object, all clients that previously did stuff like:

```rust
    // get res the usual way
    println!("Response: {}", res.status);
```

Would need to be changed to use the appropriate trait method, instead of directly accessing the field. This also implies having to import the trait itself.

Isn't this more invasive than adding a type annotation when declaring a client instance?

I have a branch [client-box-api](https://github.com/mlalic/hyper/tree/client-box-api) on my fork where you can see how it looks with this alternative approach. (It also includes the HTTP/2 commits)

seanmonstar replied:
Hm, you're right, changing that is unfortunate.

------

Here's an idea I haven't fully explored, but what I was thinking about before your pull request arrived in my inbox.

Instead of Client using a RequestFactory, I was thinking that Client would continue to use `client::Request` and `client::Response`. However, those would internally be altered to contain a sort of `HttpMessage`. I imagined it being something like this:

```rust
trait HttpMessage: Read + Write {
    fn get_incoming<T>(&mut self) -> HttpResult<http::MessageHead<T>>;
    fn set_outgoing<T>(&mut self, outgoing: http::MessageHead<T>) -> HttpResult<()>;
}
```

Request and Response would be altered:

```rust
struct Request<W> {
    message: Box<HttpMessage>,
    // ...
}

impl Request<Fresh> {
    pub fn start(self) -> HttpResult<Request<Streaming>> {
        try!(self.message.set_outgoing(MessageHead {
            subject: (self.method, self.uri),
            headers: self.headers,
            // version seems redundant?
        }));
    }
}

// and so forth...
```

The `HttpMessage` implementation would know how to set up its internal Read and Write. For H1, the implementation would adjust itself upon receiving an `Outgoing` to determine which `HttpWriter` to use.

There's currently a `http::Incoming`, that could be adjusted to work for both incoming and outgoing. Possible name is `MessageHead`.

mlalic replied:
I see. 

Essentially, we'd end up removing the (currently) private bits of the request and response structs (which are the ones that make them tightly coupled to an HTTP/1.x implementation) and replace them with this new trait object and have the structs delegate the protocol-specific operations to it. So the client keeps its api unchanged, as it still only exposes the same response struct (at least from the perspective of the public fields). 

How about we combine the two approaches... The HTTP response trait that I introduced could be replaced by a struct with the same public fields as the current one. Then, the streaming response trait mandates that the response is exactly that struct and not a generic implementation of the trait. Coupled with the boxed up factory (on the other branch that I mentioned), we'd end up removing the generic parameter and the change in the signature of the send method (at least from the api, if not binary perspective). 
I'm currently on mobile so I can't write any code snippets to get the point across better, but I hope it's understandable what I'm proposing.

What do you think about it? I might be missing something? 

(if you want to take some time to think through alternative designs yourselves, I'm fine with that and we can put this on hold for a while, I just opened the pr since I wanted to get some code out there as a basis for such a discussion as it's always easier when there's something to base it off)

seanmonstar commented on src/header/mod.rs:
````diff
@@ -164,8 +164,9 @@ impl Headers {
     /// ```
-    pub fn set_raw<K: Into<Cow<'static, str>>>(&mut self, name: K, value: Vec<Vec<u8>>) {
-        self.data.insert(UniCase(CowStr(name.into())), Item::new_raw(value));
+    pub fn set_raw<'a, K: Into<Cow<'a, str>>>(&mut self, name: K, value: Vec<Vec<u8>>) {
````
This change means the example on L165 allocates the `"content-length"` as a `String`, but there's no need.

mlalic replied:
Fair enough... I'll revert this. Come to think of it, this change isn't even required for the `http2` code anymore (as I found a different way to do what this had been intended to).

seanmonstar commented on src/http2.rs:
```diff
@@ -0,0 +1,271 @@
+            uri.push_str(&q[..]);
+        }
+
+        let mut path: Vec<u8> = Vec::new();
```
This can be `uri.into_bytes()` I believe.

mlalic replied:
Nice one!

seanmonstar commented on src/http2.rs:
```diff
@@ -0,0 +1,271 @@
+impl io::Write for Http2Request<Streaming> {
+    #[inline]
+    fn write(&mut self, msg: &[u8]) -> io::Result<usize> {
+        Ok(msg.len())
```
I think it might be better to toss an `unimplemented!()` here instead.

mlalic replied:
Sure, might make more sense for now...

seanmonstar commented on src/lib.rs:
```diff
@@ -1,6 +1,4 @@
 #![doc(html_root_url = "https://hyperium.github.io/hyper/hyper/index.html")]
-#![deny(missing_docs)]
-#![cfg_attr(test, deny(warnings))]
```
woops?

mlalic replied:
Hehe :)

As I mentioned in the opening comment, the PR is definitely not ready to be merged just yet.

I removed this since I wanted to be able to experiment with changes without being slowed down by having to put docs everywhere and get rid of warnings.

If we want to go forward with polishing this up, then of course all those issues will be solved and the commit that removed these two lines rebased out of the tree.

GitCop commented:
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 830ebe2c7c2c02b4ee7c22c3043417cafdac5a55
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 42d4364e4b2c45c8bd8e3d0905f92b26b9b9c727
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: d426c99ed4287d633978e3874a9d4b91465163cd
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: abf43e03ef1cfd6901026799db9be7dfb4bcdf88
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8e61b4351526ce468a887fc30eb0bc62f5a134a0
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 0a3c8e13a8338190c7ce16d5bb2194cf2229ee96
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4eb9786a105c74c2b940954a6f6fbf6da2c44019
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 45f509c9e9e8dd6a382a54c804e77d814488a7ba
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: acd28069931d90775605d730d23515c9945c235c
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: b3395ef688dbbb17cd8f9e3403ed5a654383abc3
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 14c1d60a2af5885cf80bcc3be8b371490f620d66
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 84925772d5411e71a29d4261d6b7dc1bcb6f65b1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ac1fe5467d2b44080c70cdcf6acea36ace4e9e3d
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 11ae013fdb5d2f82cc208a6f6ae4bd0c73fb2842
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 42bcf865f35d44876e618dece0355afa3afd60b4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4d5f293883ce65f0f6dd4cd1f995bf5b17efc590
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 9196bab0529033680b64afc0ac89ddeff24d2c98
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: d8c13fc07d86a13bfc80ddd0ed81df37276a15a1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 121e9ab84c66a2e1c0a0c70fd74503e703127545
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 6ffc23f21d1910073c1c6a24229bb82802049161
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 482e24f834434889feec8b72d2db6ada45072f10
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: e13ce02766cc3c0772ea6ca88d9f4a43f82107c6
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 190101f9f0e2f3c35d38074c542c65f91d088a08
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: f553ffddb8bbc20e0aae694cbd4934fcc32eb0b9
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4ad58bc9753e640d81d78ffc757a4193cee98ee7
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 20edf74e973ed1c382b4dfa70296ee33343061c2
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 43d3b3af56f259272a71409dad94d81c0bbb7029
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

mlalic commented on src/http2.rs:
```diff
@@ -0,0 +1,271 @@
+        // Prepare the request metadata so that it fits into `solicit`'s API.
+        // It basically just adapts the method and path into byte slices.
+        let mut method_bytes: Vec<u8> = Vec::new();
+        write!(&mut method_bytes, "{}", self.method);
```
How about we do something similar to what you suggested with using `into_bytes` for the path.

Namely, implement `Borrow<str>` for `Method` and then just pass that (`as_bytes`) to the `solicit` API that expects a `&[u8]`. It would avoid an extra allocation here...

reem replied:
Probably `AsRef<str>` would be the better thing, but sounds good.

mlalic replied:
Yeah, you're right `AsRef` would probably be better.

GitCop commented:
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 830ebe2c7c2c02b4ee7c22c3043417cafdac5a55
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 42d4364e4b2c45c8bd8e3d0905f92b26b9b9c727
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: d426c99ed4287d633978e3874a9d4b91465163cd
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: abf43e03ef1cfd6901026799db9be7dfb4bcdf88
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8e61b4351526ce468a887fc30eb0bc62f5a134a0
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 0a3c8e13a8338190c7ce16d5bb2194cf2229ee96
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4eb9786a105c74c2b940954a6f6fbf6da2c44019
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 45f509c9e9e8dd6a382a54c804e77d814488a7ba
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: acd28069931d90775605d730d23515c9945c235c
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: b3395ef688dbbb17cd8f9e3403ed5a654383abc3
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 14c1d60a2af5885cf80bcc3be8b371490f620d66
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 84925772d5411e71a29d4261d6b7dc1bcb6f65b1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ac1fe5467d2b44080c70cdcf6acea36ace4e9e3d
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 11ae013fdb5d2f82cc208a6f6ae4bd0c73fb2842
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 42bcf865f35d44876e618dece0355afa3afd60b4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4d5f293883ce65f0f6dd4cd1f995bf5b17efc590
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 9196bab0529033680b64afc0ac89ddeff24d2c98
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: d8c13fc07d86a13bfc80ddd0ed81df37276a15a1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 121e9ab84c66a2e1c0a0c70fd74503e703127545
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 6ffc23f21d1910073c1c6a24229bb82802049161
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 482e24f834434889feec8b72d2db6ada45072f10
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: e13ce02766cc3c0772ea6ca88d9f4a43f82107c6
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 190101f9f0e2f3c35d38074c542c65f91d088a08
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: f553ffddb8bbc20e0aae694cbd4934fcc32eb0b9
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4ad58bc9753e640d81d78ffc757a4193cee98ee7
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 20edf74e973ed1c382b4dfa70296ee33343061c2
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 43d3b3af56f259272a71409dad94d81c0bbb7029
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 625a562abb4006e893960170eeae628fb3347960
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 830ebe2c7c2c02b4ee7c22c3043417cafdac5a55
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 42d4364e4b2c45c8bd8e3d0905f92b26b9b9c727
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: d426c99ed4287d633978e3874a9d4b91465163cd
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: abf43e03ef1cfd6901026799db9be7dfb4bcdf88
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8e61b4351526ce468a887fc30eb0bc62f5a134a0
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 0a3c8e13a8338190c7ce16d5bb2194cf2229ee96
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4eb9786a105c74c2b940954a6f6fbf6da2c44019
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 45f509c9e9e8dd6a382a54c804e77d814488a7ba
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: acd28069931d90775605d730d23515c9945c235c
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: b3395ef688dbbb17cd8f9e3403ed5a654383abc3
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 14c1d60a2af5885cf80bcc3be8b371490f620d66
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 84925772d5411e71a29d4261d6b7dc1bcb6f65b1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ac1fe5467d2b44080c70cdcf6acea36ace4e9e3d
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 11ae013fdb5d2f82cc208a6f6ae4bd0c73fb2842
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 42bcf865f35d44876e618dece0355afa3afd60b4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4d5f293883ce65f0f6dd4cd1f995bf5b17efc590
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 9196bab0529033680b64afc0ac89ddeff24d2c98
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: d8c13fc07d86a13bfc80ddd0ed81df37276a15a1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 121e9ab84c66a2e1c0a0c70fd74503e703127545
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 6ffc23f21d1910073c1c6a24229bb82802049161
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 482e24f834434889feec8b72d2db6ada45072f10
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: e13ce02766cc3c0772ea6ca88d9f4a43f82107c6
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 190101f9f0e2f3c35d38074c542c65f91d088a08
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: f553ffddb8bbc20e0aae694cbd4934fcc32eb0b9
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4ad58bc9753e640d81d78ffc757a4193cee98ee7
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 20edf74e973ed1c382b4dfa70296ee33343061c2
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 43d3b3af56f259272a71409dad94d81c0bbb7029
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 625a562abb4006e893960170eeae628fb3347960
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 87132f8e94f08c508361c2305b9fcaed18a6ed61
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: cb39b60812b209a0b83018eff0e48802de40b06f
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

boghison commented:
+1, been waiting for this ever since HTTP2 came out :)

reem commented:
@mlalic might be worth rewriting commits now just to get gitcop to shut up.

reem commented on src/client/mod.rs:
```diff
@@ -32,39 +32,120 @@ use status::StatusClass::Redirection;
+    /// `new()` call. This, however, means that all particular
+    /// `HttpRequestFactory` implementations must have a set of default
+    /// parameters with which they can be created.
+    fn new() -> Self;
```
We could get rid of this and just have `HttpRequestFactory` inherit from `Default`.

mlalic replied:
Nice one.

reem commented on src/client/mod.rs:
```diff
@@ -32,39 +32,120 @@ use status::StatusClass::Redirection;
+/// particular `FreshHttpRequest` starting point.
+///
+/// TODO: Might need a better name?
+pub trait HttpRequestFactory {
```
One possible renaming: the trait can be called `HttpProtocol` or just `Protocol` and the types can just be renamed to `Http11` and `Http2`.

mlalic replied:
It sounds good. My feeling is that it might imply a bit more than what the trait currently does (only deals with the client side aspect), but bikeshedding over the name too much doesn't gain us anything...

reem replied:
Since this is in the client module already, I'm not too concerned with it being overly broad, and I'd be really happy to get rid of `HttpRequestBoxFactory` :P

GitCop commented:
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 3ce492a13535b223636212eb59c7f08995b0a791
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 3ce492a13535b223636212eb59c7f08995b0a791
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

mlalic commented:
So I've updated the PR based on the discussion so far: rebased the fixes that were made along the way to the corresponding commits and adapted the commit messages to the style guidelines (except the one that needs to be rebased out of the tree anyway!).

The status is that with this, the public API of the client seems to be *very nearly* unaffected.

The return type is now `HttpResult<HttpResponse>`, where `HttpResponse` is a new struct that has the same public API as the previous `Response` struct, *apart from* not being generic. Therefore, all the examples and tests can work without any changes, but if anyone relied on the fact that this public type is generic, that code would break. Obviously, the name of the type is also different, but we could easily call this one `Response` and rename the old one, although in that case a different type of "breakage" occurs, since `Response` was a publicly exposed type, someone might have used it directly, instead of going through the `Client` (though very unlikely).

For now I also removed the `with_connector` method, but we could keep that in and instantiate an HTTP/1.1 request factory in the implementation so as to keep that change also completely transparent.

mlalic commented:
So, what do you guys (@seanmonstar, @reem) think: should we go down this route? If so, I can bang this into shape with proper test coverage, documentation, error handling, etc.

seanmonstar commented:
@mlalic sorry, been busy. I'll look it over again today.

seanmonstar commented:
@mlalic I know Servo uses `Request::with_connector`, cause the XHR spec defines some fun rules slightly different than standard HTTP. /cc @Manishearth 

I have a Pool implementation in #486. That would allow keeping the requests around, and having h1 -> h2 upgrade functionality.

What did you think of my [previous suggestion](https://github.com/hyperium/hyper/pull/448#discussion_r28260410) of keeping Client, Request, and Response as opaque structs, and having the trait be a `HttpMessage`, implemented by http1 and http2? I'd like to think this would ease implementing on the server as well...

mlalic commented:
@seanmonstar 

> I know Servo uses Request::with_connector, cause the XHR spec defines some fun rules slightly different than standard HTTP

Indeed, currently the establishment of the underlying TCP connection that the HTTP/2 connection would use is somewhat hidden away, but letting it use the `NetworkConnector` abstraction for that should be quite simple. The `NetworkConnector` produces an object that is both `Read` and `Write`, which is all that `solicit`'s HTTP/2 connection requires. The custom connectors that Servo may already have should then work regardless of whether the tcp connection is used for h1 or h2.

> I have a Pool implementation in #486. That would allow keeping the requests around, and having h1 -> h2 upgrade functionality.

The way the pool is implemented right now doesn't lend itself at all to reusing connections for HTTP/2. What is currently pooled are the tcp connections, not HTTP connections. This is virtually the same for h1, but it is not so for h2. This is because in h1, you can simply send the raw request bytes on one of the established tcp streams and you're good; no previously sent requests on that particular stream matter for the raw payload that you transmit over the wire, nor for the parsing of the response payload. In h2, however, the payload you send for a request does depend on previously sent requests - you need to assign a new stream id, hpack encoding/decoding the headers is stateful, flow control windows depend on previously sent data, etc. Therefore, simply pooling the TCP connections does nothing to allow "keep alive"-like usage of an HTTP/2 connection.

Of course, it's doable to pool the actual HTTP 2 connections, by doing what you've done on the connector level within (what is currently named in this pr) the `HttpRequestFactory`. In fact, if acquiring a lock before creating a new request is already deemed acceptable, I could even squeeze my async client into the same api, allowing multiple concurrently issued requests to the same host to use the same HTTP 2 connection, as long as the original factory instance is kept alive and its clones given to new hyper clients (provided that spawning an io thread is also acceptable).

> What did you think of my previous suggestion of keeping Client, Request, and Response as opaque structs, and having the trait be a HttpMessage, implemented by http1 and http2? I'd like to think this would ease implementing on the server as well...

I don’t think I fully understood what you were aiming at with the `HttpMessage` trait. From what I can gather, it should be a type that exposes an interface for queuing HTTP "messages" (`set_outgoing`), as well as for grabbing them off a connection (`get_incoming`). The `Read`/`Write` trait bounds were for reading/writing the body?

The way it looks like it would be used would to implement the request's `Fresh -> Streaming -> Response` transformation would be to first queue the `MessageHead` (i.e. `Fresh -> Streaming`), then stream out the response (calls to the underlying `Write`?) and grab the response head (i.e. `Streaming -> Response`), and finally read out the response (calls to the `Read`). All this has already become encapsulated by the `FreshHttpRequest` and `StreamingHttpRequest` traits. The implementations of these two traits, along with a corresponding factory for HTTP/1.1 and HTTP/2 fully encapsulate all relevant protocol logic.

So, I don’t see the benefit of this alternate approach. It seems it would only shift the actual protocol logic a level lower... Is it that you hope to reuse this `HttpMessage` component to support server-side HTTP/2, as well?

Originally, when you wrote that comment, I thought the suggestion was so that we avoid having type parameters on the `Client`, which I’ve achieved by making the `Response` struct fully independent from the protocol that produces it. (The original post is what nudged me into realizing it was easily doable.)

Perhaps you could flesh out your proposal more because I just might not be fully getting the idea from the previous short description.

seanmonstar commented:
I'll expore in a branch to show what I mean.

mlalic commented:
@seanmonstar I have an experiment over on [this branch](https://github.com/hyperium/hyper/compare/master...mlalic:http-message) (forked off the current master HEAD, not this PR) refactoring the HTTP/1.1-specific parts of `Request`/`Response` into an implementation of an `HttpMessage` trait, such as what you described.

Is this sort of what you had in mind for the `HttpMessage` API?

seanmonstar commented:
Pretty much exactly. As long as such a trait could work for http2, of
course.

On Mon, May 25, 2015, 3:56 PM Marko Lalic <notifications@github.com> wrote:

> @seanmonstar <https://github.com/seanmonstar> I have an experiment over
> on this branch
> <https://github.com/hyperium/hyper/compare/master...mlalic:http-message>
> (forked off the current master HEAD, not this PR) refactoring the
> HTTP/1.1-specific parts of Request/Response into an implementation of an
> HttpMessage trait, such as what you described.
>
> Is this sort of what you had in mind for the HttpMessage API?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hyperium/hyper/pull/448#issuecomment-105331881>.
>

mlalic commented:
@seanmonstar 

The trait does work out quite nicely for HTTP/2 too.

So, here's an attempt at landing the PR, this time with everything polished up with proper docs and tests for the new features. Feedback of course appreciated.

(Side note: the `HttpMessage` API would allow for some nicer tests for the `Client` itself. Right now, it seems very difficult -- or at least extremely kludgy -- to test what the client itself sends, regardless of the protocol. For example, I wanted to write a test which checks whether the client properly handles redirects, as by looking at the code it seems like it always sticks to the original request method, regardless of the redirect status code it sees, but I gave up once I saw that writing a test that checks exactly what requests got sent wasn't as easy as I'd hoped. With the `HttpMessage` API we can just test which messages got dispatched, making it all much smoother.)

seanmonstar commented:
Sorry, I was camping this weekend (so no Internet), but this looks exciting!

seanmonstar commented on src/error.rs:
```diff
@@ -108,6 +114,12 @@ impl From<httparse::Error> for Error {
 
+impl From<Http2Error> for Error {
+    fn from(err: Http2Error) -> Error {
+        Error::Http2(err)
```
Could you add a `from_and_cause!` case to the tests below, to keep the coverage happy.

mlalic replied:
Done.

seanmonstar commented on src/http11.rs:
```diff
@@ -0,0 +1,311 @@
+        let raw_status = head.subject;
+        let headers = head.headers;
+
+        /*
```
Was this commented out originally? Seems it can be killed.

mlalic replied:
Indeed, this is still done in the `Response` so as not to duplicate it in all `HttpMessage` impls. Removed.

seanmonstar commented on src/http11.rs:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    fn close_connection(&mut self) -> ::Result<()> {
+        try!(self.get_mut().close(Shutdown::Both));
```
This could just be `self.get_mut().close(...)`, right?

mlalic replied:
If you mean without the `try!`, then no. The error types are mismatched: `close_connection` requires `hyper::error::Error` and the `close` returns an `std::io::Error`. Since the types aren't automatically converted (i.e. without a `From::from` call), to me it looked nicest to go with a `try!`.

seanmonstar commented on src/http11.rs:
```diff
@@ -0,0 +1,311 @@
+/// An implementation of the `HttpMessage` trait for HTTP/1.1.
+#[derive(Debug)]
+pub struct Http11Message {
+    stream: Option<Box<NetworkStream + Send>>,
```
Instead of having 3 options (and the possible panic! states below if all are `None`), how about some sort of `MessageState` enum, with the 3 variants. I think that would be clearer that these determine the state of the message, whereas reading the 3 options at first left me confused. Additionally, we could remove the `panic!`s where all 3 options are None, since the enum will enforce for us the state.

mlalic replied:
All right, say there's an enum:

```rust
enum MessageState {
    Idle(Box<NetworkStream + Send>),
    Writing(HttpWriter<BufWriter<Box<NetworkStream + Send>>>),
    Reading(HttpReader<BufReader<Box<NetworkStream + Send>>>),
}
```

Now in `set_outgoing(&mut self, ...)`, the internal state of the message needs to transition to `MessageState::Writing` and the writer needs to be based on the stream originally wrapped in the `Idle` variant. I'm not sure how the `Box<NetworkStream>` should be moved into a different variant here.

`set_outgoing` takes `&mut self`, so neither the full state enum nor the `Box` within the `Idle` variant can be moved. The only thing I can think of around that would be to use `mem::replace`, but it'd still require something (akin to `None`) to be put in place of the original state while the stream is juggled into the correct Writer.

This `None`-like state, though, would still imply that it's technically possible to lose the stream along the way, so it's only a slight improvement...

Is there some better approach to implementing the transition which would avoid this `None` state altogether?

seanmonstar replied:
Fair enough. It could probably be done inside a method on the state enum, but don't worry about it. I might look at if it's possible to keep state similar to the `WriteStatus` that `Request` uses.

seanmonstar commented on src/http2.rs:
```diff
@@ -0,0 +1,563 @@
+/// the `hyper::client` module.
+pub struct Http2Message<S> where S: CloneableStream {
+    client: SimpleClient<Http2Stream<S>>,
+    request: Option<Http2Request>,
```
Probably similar here: if these options are for tracking state, a `MessageState` enum would probably be clearer.

mlalic replied:
Here, on the other hand, the state enum works out just fine. I suppose because there's no state that needs to be transferred between the variants when a transition happens...

seanmonstar commented on src/http2.rs:
```diff
@@ -0,0 +1,563 @@
+            // Unfortunately, `hyper` wants its header names to be strings...
+            // We just panic if it ends up being an invalid utf8 string, but no
+            // copies are made.
+            headers.set_raw(String::from_utf8(name).unwrap(), vec![value]);
```
The spec says that header names are valid tokens, which means they should be valid utf8. Is that not the case in http2? Either way, I'm not sure an `unwrap` should be used.

Also, in Http11, the `Set-Cookie` header is sent multiple times if there are multiple cookies. Is that different in Http2?

mlalic replied:
The names should still be valid utf8 ([8.1.2 HTTP Header Fields](http://http2.github.io/http2-spec/#HttpHeaders)) on the HTTP layer, that's true. (HPACK is what treats both the name and value as an opaque octet sequence.) So, indeed, if the name cannot be decoded as utf8, the response is malformed. This is now fixed (along with a test).

As for multiple `SetCookie` headers -- the situation is unchanged: multiple cookies correspond to multiple `set-cookies` headers. That said, I was hoping that `hyper` would correctly handle this itself...

What I see now, though, is that it effectively fakes it by having the *value* of the `Set-Cookie` entry in the `header::Headers` struct contain a CRLF delimiter and continue on to the next `Set-Cookie` header.

```rust
    let cookies = header::SetCookie(vec![
        cookie::Cookie::new("foo".to_owned(), "bar".to_owned()),
        cookie::Cookie::new("baz".to_owned(), "quux".to_owned())
    ]);
    // Now the following is true:
    cookies.name() == "Set-Cookie"
    cookies.value_string() == "foo=bar; Path=/\r\nSet-Cookie: baz=quux; Path=/"
```

That hack works when sending the header over HTTP/1.1, but not HTTP/2, as the CRLF is meaningless. I've worked around it in the `prepare_headers` function to special case the `Set-Cookie` and extract the cookies that it contains into separate HTTP/2 headers (although since for now there's only client support, it shouldn't actually get triggered, as clients aren't supposed to send that header).

On the other hand, when parsing a response with multiple `set-cookie` headers returned by an HTTP/2 server, it seems that `hyper`'s `Headers::from_raw` method ends up returning a headers collection with a `SetCookie` header that correctly contains all the cookies, as verified by a new unit test... Should `from_raw` not be relied on to do the right thing?

seanmonstar replied:
`Headers::from_raw` will do the right thing. However, `headers.set_raw()` will *override* whatever value was there before, like any setter.

The cookie formatting hack was indeed a nasty hack. We've been trying to think of a way to allow headers to indicate they output over multiple lines, but keeping that invisible from 99% of headers (since they don't need it) is difficult. I'd like to try to figure that out, such that a `Header` can return a `Encoder` that implements `Iterator`, and also be able to pass that to solicit, such that the bytes are written only once through hpack, instead of allocating a temporary vector beforehand.

mlalic replied:
That would be a nice improvement to aim for, for sure.

seanmonstar commented:
Just curious, from the example: `"Host": "http2bin.org,http2bin.org",`, is it supposed to have the host doubled like that?

seanmonstar commented:
@mlalic woo! thanks for this herculean effort!

mlalic commented:
Happy to do it! Glad we've worked out the initial integration; we can take it from here and tweak/improve things incrementally, of course... Should be less of a big-bang than this :)

mlalic commented:
> Just curious, from the example: `"Host": "http2bin.org,http2bin.org",`, is it supposed to have the host doubled like that?

`hyper` probably includes the `Host` header and the HTTP/2 client includes the `:authority` pseudo-header so that's probably where it ends up looking doubled.

`:authority` "should" be used in HTTP/2 requests, but `host` is still fine, so we can tweak it that the `Http2Message` doesn't set any of those unless they're given to it in the message head? (Like HTTP/1.1 at least one of those headers is still required for every request, so it means that the `Request` always needs to set it...)
